src/vmm/src/serial.rs

```diff
@@ -14,42 +14,78 @@ use vm_device::{
+                    Err(e) => {
+                        eprintln!("Error while reading stdin: {:?}", e);
+                    }
+                    Ok(count) => {
+                        let event_set = events.event_set();
+                        let unregister_condition = event_set.contains(EventSet::ERROR)
+                            | event_set.contains(EventSet::HANG_UP);
+                        if count > 0 {
+                            if self.0.enqueue_raw_bytes(&out[..count]).is_err() {
+                                eprintln!("Failed to send bytes to the guest via serial input");
+                            }
+                        } else if unregister_condition {
+                            // Got 0 bytes from serial input; is it a hang-up or error?
+                            ops.remove(events)
+                                .expect("Failed to unregister serial input");
+                        }
+                    }
+                }
             }
-            Ok(count) => {
-                let event_set = events.event_set();
-                let unregister_condition =
-                    event_set.contains(EventSet::ERROR) | event_set.contains(EventSet::HANG_UP);
-                if count > 0 {
-                    if self.0.enqueue_raw_bytes(&out[..count]).is_err() {
-                        eprintln!("Failed to send bytes to the guest via serial input");
+            None => {
+                match stdin().read(&mut out) {
+                    Err(e) => {
+                        eprintln!("Error while reading stdin: {:?}", e);
+                    }
+                    Ok(count) => {
+                        let event_set = events.event_set();
+                        let unregister_condition = event_set.contains(EventSet::ERROR)
+                            | event_set.contains(EventSet::HANG_UP);
+                        if count > 0 {
+                            if self.0.enqueue_raw_bytes(&out[..count]).is_err() {
+                                eprintln!("Failed to send bytes to the guest via serial input");
+                            }
+                        } else if unregister_condition {
+                            // Got 0 bytes from serial input; is it a hang-up or error?
+                            ops.remove(events)
+                                .expect("Failed to unregister serial input");
+                        }
                     }
-                } else if unregister_condition {
-                    // Got 0 bytes from serial input; is it a hang-up or error?
-                    ops.remove(events)
-                        .expect("Failed to unregister serial input");
                 }
             }
-        }
+        };
     }
 
     fn init(&mut self, ops: &mut EventOps) {
         // Hook to stdin events.
-        ops.add(Events::new(&stdin(), EventSet::IN))
-            .expect("Failed to register serial input event");
+        match &self.1 {
```

Comment:
This would also be collapsed into a single arm if the `Read`er part was generic, I believe.